retrieval/discovery/serverset.go

```diff
@@ -327,12 +329,13 @@ func (tc *zookeeperTreeCache) recursiveNodeUpdate(path string, node *zookeeperTr
 				children: map[string]*zookeeperTreeCacheNode{},
 				done:     make(chan struct{}, 1),
 			}
-		}
-		err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
-		if err != nil {
-			return err
+			err = tc.recursiveNodeUpdate(path+"/"+child, node.children[child])
```

Comment:
http://www.pongasoft.com/blog/yan/entry/zookeeper_loss_of_events_problem/ was the page I was thinking of, it seems like this will work okay as the watch on the child will catch it changing.